### PR TITLE
Update SMOS L2 SM DA routine

### PR DESCRIPTION
**Pull request by smodanesi**
### PR for the update of the SMOS L2 DA into LIS.

This PR encompasses several enhancements:
- changes to enable the SMOS L2 soil moisture DA into LIS
- an update of the main functions for reading/writing the SMOS L2 soil moisture observations in the DA tool, including the application of the recommended confidence and science flags
- implementation of capability to read SMOS L2 soil moisture from netCDF files, replacing the previous .DBL file format. 
- connection between the SMOS L2 soil moisture DA functions and the Noah-MP3.6 and Noah-MP4.0.1 LSMs

Note that the CDF-scaling approach of SMOS L2 soil moisture to the LSM climatology has not been tested yet (see in lis/dataassim/obs/SMOS_L2sm/read_SMOSL2sm.F90). The system was tested for unscaled observation (i.e., no bias correction).

### CODE REVIEW

**Review by mbechtold — APPROVED** [review]
Hi Sara, many thx.
It seems you eventually don't use LIS_ensOnGrid, so no need to add it to LIS_coreMod.F90.
The rest looks good to me. I checked the compilation, it worked for me too.

**Comment by smodanesi** [comment]
Hi Michel, as you suggested I have removed the changes related to LIS_ensOnGrid in LIS_coreMod.F90. I chacked the compilation and it worked for me. Thank you too and let me know if there is anything to change.

**Comment by mbechtold** [comment]
Hi Sara, for me this is good now. Many thanks.

File: lis/dataassim/obs/SMOS_L2sm/SMOSL2sm_Mod.F90 [hunk 1]
```diff
@@ -205,9 +207,10 @@ subroutine SMOSL2sm_setup(k, OBS_State, OBS_Pert_State)
        call ESMF_AttributeSet(OBS_State(n),"Data Assimilate Status",&
             .false., rc=status)
        call LIS_verify(status)
-       
+
+      !SM09012024 change LIS_rc%ngrid(n) with LIS_rc%obs_ngrid(k) 
        call ESMF_AttributeSet(OBS_State(n),"Number Of Observations",&
-            LIS_rc%ngrid(n),rc=status)
+            LIS_rc%obs_ngrid(k),rc=status)
        call LIS_verify(status)
```
Comment (gdelannoy) [inline lis/dataassim/obs/SMOS_L2sm/SMOSL2sm_Mod.F90:214]: I don't understand exactly what this is doing: there is a loop over n->nnest, and obs_ngrid(k) is overwritten for all nnest?

**Review by gdelannoy — COMMENTED** [review]
Will reach out to you.

File: lis/dataassim/obs/SMOS_L2sm/SMOSL2sm_Mod.F90 (hunk 1 above)
Comment (gdelannoy) [reply lis/dataassim/obs/SMOS_L2sm/SMOSL2sm_Mod.F90:214]: (k) is assimilation instance; in line with SMAP DA. The old SMOS DA code was not functioning.

**Review by gdelannoy — APPROVED** [review]
Very well done, Sara, thanks!

**Comment by smodanesi** [comment]
Dear Gabrielle, many thanks!

**Pull request merged**